File: generation-pythia/scripts/generate_minbias.py
#!/usr/bin/env python3
import argparse
import os
from pathlib import Path
import sys

import pyarrow as pa
import pyarrow.parquet as pq
import pythia8mc as pythia8

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from common.config_utils import resolve_minbias_campaign  # noqa: E402


def parse_args():
    parser = argparse.ArgumentParser(
        description="Generate Pythia8 minbias events and write a flat particle parquet file."
    )
    parser.add_argument("--events", type=int, default=1000, help="Number of accepted events")
    parser.add_argument("--e-cm", type=float, default=14000.0, help="Center-of-mass energy in GeV")
    parser.add_argument(
        "--processes",
        default="SoftQCD:all",
        help='Pythia8 process switch to enable, e.g. "SoftQCD:all"',
    )
    parser.add_argument("--seed", type=int, default=None, help="Optional Pythia random seed")
    parser.add_argument(
        "--campaign",
        "--out-tag",
        dest="campaign",
        default="minbias",
        help="Campaign name for the default output path",
    )
    parser.add_argument("-o", "--output", default=None, help="Output parquet file")
    parser.add_argument("--verbose", action="store_true", help="Print event progress")
    args = parser.parse_args()

    if args.events <= 0:
        raise ValueError("--events must be > 0")
    if args.e_cm <= 0.0:
        raise ValueError("--e-cm must be > 0")
    if args.seed is not None and args.seed < 0:
        raise ValueError("--seed must be non-negative")
    if not args.campaign:
        raise ValueError("--campaign must not be empty")
    if args.output is None:
        campaign_dir, campaign = resolve_minbias_campaign(args.campaign)
        args.output = campaign_dir / f"{campaign}.parquet"
    else:
        args.output = Path(args.output)
    return args


def configure_pythia(e_cm, processes, seed=None, verbose=False):
    os.environ.pop("PYTHIA8DATA", None)

    pythia = pythia8.Pythia("", verbose)
    pythia.readString("Beams:idA = 2212")
    pythia.readString("Beams:idB = 2212")
    pythia.readString(f"Beams:eCM = {e_cm}")

    pythia.readString("SoftQCD:nonDiffractive      = off")
    pythia.readString("SoftQCD:elastic             = off")
    pythia.readString("SoftQCD:singleDiffractive   = off")
    pythia.readString("SoftQCD:doubleDiffractive   = off")
    pythia.readString("SoftQCD:centralDiffractive  = off")
    pythia.readString(f"{processes} = on")

    pythia.readString("Next:numberShowInfo = 0")
    pythia.readString("Next:numberShowProcess = 0")
    pythia.readString("Next:numberShowEvent = 0")
    if not verbose:
        pythia.readString("Print:quiet = on")
        pythia.readString("Init:showProcesses = off")
        pythia.readString("Init:showChangedSettings = off")
        pythia.readString("Init:showChangedParticleData = off")
    if seed is not None:
        pythia.readString("Random:setSeed = on")
        pythia.readString(f"Random:seed = {seed}")

    if not pythia.init():
        raise RuntimeError("Pythia initialization failed")
    return pythia


def empty_columns():
    return {
        "event_id": [],
        "particle_index": [],
        "pdg_id": [],
        "status": [],
        "is_final": [],
        "charge": [],
        "px": [],
        "py": [],
        "pz": [],
        "E": [],
        "m": [],
        "pt": [],
        "eta": [],
        "phi": [],
    }


def append_event(columns, event_id, event):
    for particle_index in range(1, event.size()):
        particle = event[particle_index]
        columns["event_id"].append(event_id)
        columns["particle_index"].append(particle_index)
        columns["pdg_id"].append(int(particle.id()))
        columns["status"].append(int(particle.status()))
        columns["is_final"].append(bool(particle.isFinal()))
        columns["charge"].append(float(particle.charge()))
        columns["px"].append(float(particle.px()))
        columns["py"].append(float(particle.py()))
        columns["pz"].append(float(particle.pz()))
        columns["E"].append(float(particle.e()))
        columns["m"].append(float(particle.m()))
        columns["pt"].append(float(particle.pT()))
        columns["eta"].append(float(particle.eta()))
        columns["phi"].append(float(particle.phi()))


def write_parquet(path, columns):
    schema = pa.schema(
        [
            ("event_id", pa.int64()),
            ("particle_index", pa.int32()),
            ("pdg_id", pa.int32()),
            ("status", pa.int32()),
            ("is_final", pa.bool_()),
            ("charge", pa.float64()),
            ("px", pa.float64()),
            ("py", pa.float64()),
            ("pz", pa.float64()),
            ("E", pa.float64()),
            ("m", pa.float64()),
            ("pt", pa.float64()),
            ("eta", pa.float64()),
            ("phi", pa.float64()),
        ]
    )
    table = pa.Table.from_pydict(columns, schema=schema)
    path.parent.mkdir(parents=True, exist_ok=True)
    pq.write_table(table, path)


def generate(args):
    pythia = configure_pythia(args.e_cm, args.processes, args.seed, args.verbose)
    columns = empty_columns()

    written = 0
    consecutive_failures = 0
    max_failures = 1000
    while written < args.events:
        if not pythia.next():
            consecutive_failures += 1
            if consecutive_failures >= max_failures:
                raise RuntimeError("Too many consecutive Pythia event failures")
            continue

        consecutive_failures = 0
        written += 1
        append_event(columns, written, pythia.event)
        if args.verbose and (written <= 5 or written % 1000 == 0):
            print(f"Event {written}: pythia_entries={pythia.event.size()}")

    write_parquet(args.output, columns)
    return written, len(columns["event_id"])


def main():
    try:
        args = parse_args()
        events, particles = generate(args)
        print(f"Wrote {particles} particles from {events} events to {args.output}")
    except Exception as exc:
        raise SystemExit(f"ERROR: {exc}") from None


if __name__ == "__main__":
    main()

File: generation-pythia/scripts/generate_minbias.cc
#include <Pythia8/Pythia.h>
#include <Pythia8Plugins/HepMC3.h>

#include <HepMC3/GenEvent.h>
#include <HepMC3/Units.h>
#include <HepMC3/WriterAscii.h>

#include <cstdlib>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

namespace fs = std::filesystem;

struct Args {
  long events = 1000;
  double e_cm = 14000.0;
  std::string processes = "SoftQCD:all";
  int seed = -1;
  std::string campaign = "minbias";
  fs::path output;
  bool verbose = false;
};

bool looks_like_repo_root(const fs::path& path) {
  return fs::exists(path / "setup_env.sh") && fs::exists(path / "generation-pythia");
}

fs::path find_repo_root(fs::path start) {
  start = fs::absolute(start);
  if (fs::is_regular_file(start)) {
    start = start.parent_path();
  }

  for (fs::path path = start; !path.empty(); path = path.parent_path()) {
    if (looks_like_repo_root(path)) {
      return path;
    }
    if (path == path.root_path()) {
      break;
    }
  }
  return {};
}

fs::path repo_root(const char* argv0) {
  if (const char* env = std::getenv("HIGGS_CEP_STUDIES_DIR")) {
    return fs::path(env);
  }

  fs::path root = find_repo_root(argv0);
  if (!root.empty()) {
    return root;
  }
  root = find_repo_root(fs::current_path());
  if (!root.empty()) {
    return root;
  }
  return fs::current_path();
}

void usage(const char* argv0) {
  std::cerr
      << "Usage:\n"
      << "  " << argv0 << " [--events N] [--e-cm ECM_GEV]\n"
      << "       [--processes SoftQCD:all] [--seed SEED]\n"
      << "       [--campaign NAME|--out-tag NAME] [--output OUTPUT.hepmc]\n"
      << "       [--verbose]\n";
}

Args parse_args(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    auto require_value = [&](const std::string& name) -> std::string {
      if (i + 1 >= argc) {
        throw std::runtime_error("Missing value for " + name);
      }
      return argv[++i];
    };

    if (arg == "--events") {
      args.events = std::stol(require_value(arg));
    } else if (arg == "--e-cm") {
      args.e_cm = std::stod(require_value(arg));
    } else if (arg == "--processes") {
      args.processes = require_value(arg);
    } else if (arg == "--seed") {
      args.seed = std::stoi(require_value(arg));
    } else if (arg == "--campaign" || arg == "--out-tag") {
      args.campaign = require_value(arg);
    } else if (arg == "--output" || arg == "-o") {
      args.output = require_value(arg);
    } else if (arg == "--verbose") {
      args.verbose = true;
    } else if (arg == "--help" || arg == "-h") {
      usage(argv[0]);
      std::exit(0);
    } else {
      throw std::runtime_error("Unknown argument: " + arg);
    }
  }

  if (args.events <= 0) {
    throw std::runtime_error("--events must be > 0");
  }
  if (args.e_cm <= 0.0) {
    throw std::runtime_error("--e-cm must be > 0");
  }
  if (args.seed < -1) {
    throw std::runtime_error("--seed must be non-negative");
  }
  if (args.campaign.empty()) {
    throw std::runtime_error("--campaign must not be empty");
  }
  if (args.output.empty()) {
    args.output = repo_root(argv[0]) / "output" / "minbias" / args.campaign /
                  (args.campaign + ".hepmc");
  }

  return args;
}

void configure_pythia(Pythia8::Pythia& pythia, const Args& args) {
  pythia.readString("Beams:idA = 2212");
  pythia.readString("Beams:idB = 2212");
  pythia.readString("Beams:eCM = " + std::to_string(args.e_cm));

  pythia.readString("SoftQCD:nonDiffractive      = off");
  pythia.readString("SoftQCD:elastic             = off");
  pythia.readString("SoftQCD:singleDiffractive   = off");
  pythia.readString("SoftQCD:doubleDiffractive   = off");
  pythia.readString("SoftQCD:centralDiffractive  = off");
  pythia.readString(args.processes + " = on");

  pythia.readString("Next:numberShowInfo = 0");
  pythia.readString("Next:numberShowProcess = 0");
  pythia.readString("Next:numberShowEvent = 0");
  if (!args.verbose) {
    pythia.readString("Print:quiet = on");
    pythia.readString("Init:showProcesses = off");
    pythia.readString("Init:showChangedSettings = off");
    pythia.readString("Init:showChangedParticleData = off");
  }
  if (args.seed >= 0) {
    pythia.readString("Random:setSeed = on");
    pythia.readString("Random:seed = " + std::to_string(args.seed));
  }
}

void write_event(Pythia8::Pythia& pythia,
                 HepMC3::Pythia8ToHepMC3& converter,
                 HepMC3::WriterAscii& writer,
                 long event_number) {
  HepMC3::GenEvent event(HepMC3::Units::GEV, HepMC3::Units::MM);
  if (!converter.fill_next_event(pythia, event)) {
    throw std::runtime_error("HepMC conversion failed for event " + std::to_string(event_number));
  }
  event.set_event_number(static_cast<int>(event_number));
  writer.write_event(event);
  if (writer.failed()) {
    throw std::runtime_error("HepMC writing failed for event " + std::to_string(event_number));
  }
}

int main(int argc, char** argv) {
  try {
    const Args args = parse_args(argc, argv);

    if (args.output.has_parent_path()) {
      fs::create_directories(args.output.parent_path());
    }

    Pythia8::Pythia pythia("", args.verbose);
    configure_pythia(pythia, args);
    if (!pythia.init()) {
      throw std::runtime_error("Pythia initialization failed");
    }

    HepMC3::WriterAscii writer(args.output.string());
    if (writer.failed()) {
      throw std::runtime_error("Could not create output file: " + args.output.string());
    }
    HepMC3::Pythia8ToHepMC3 converter;

    long written = 0;
    int consecutive_failures = 0;
    constexpr int max_failures = 1000;
    while (written < args.events) {
      if (!pythia.next()) {
        ++consecutive_failures;
        if (consecutive_failures >= max_failures) {
          throw std::runtime_error("Too many consecutive Pythia event failures");
        }
        continue;
      }

      consecutive_failures = 0;
      ++written;
      write_event(pythia, converter, writer, written);
      if (args.verbose && (written <= 5 || written % 1000 == 0)) {
        std::cout << "Event " << written << ": pythia_entries=" << pythia.event.size()
                  << " weight=" << pythia.info.weight() << "\n";
      }
    }

    writer.close();
    if (writer.failed()) {
      throw std::runtime_error("HepMC writing failed while closing " + args.output.string());
    }

    std::cout << "Wrote " << written << " events to " << args.output << "\n";
    return 0;
  } catch (const std::exception& err) {
    std::cerr << "ERROR: " << err.what() << "\n";
    usage(argv[0]);
    return 1;
  }
}
